drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1082 @@
+
+	if (set_crtc_property(req, &dev->crtc, dev->crtc_id, "ACTIVE", 1) < 0)
+		return -1;
+
+	flags = DRM_MODE_ATOMIC_ALLOW_MODESET;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+static int modeset_atomic_commit(int fd, struct modeset_buf *buf, struct modeset_dev *dev)
+{
+	struct drm_object *plane = &dev->plane;
+        drmModeAtomicReq *req;
+	int ret, flags;
+
+        req = drmModeAtomicAlloc();
+
+        set_plane_property(req, plane, dev->plane_id, "FB_ID", buf->fb);
+        set_plane_property(req, plane, dev->plane_id, "CRTC_ID", dev->crtc_id);
+        set_plane_property(req, plane, dev->plane_id, "SRC_X", 0);
+        set_plane_property(req, plane, dev->plane_id, "SRC_Y", 0);
+        set_plane_property(req, plane, dev->plane_id, "SRC_W", buf->width << 16);
+        set_plane_property(req, plane, dev->plane_id, "SRC_H", buf->height << 16);
+        set_plane_property(req, plane, dev->plane_id, "CRTC_X", 0);
+        set_plane_property(req, plane, dev->plane_id, "CRTC_Y", 0);
+        set_plane_property(req, plane, dev->plane_id, "CRTC_W", buf->width);
+        set_plane_property(req, plane, dev->plane_id, "CRTC_H", buf->height);
+
+	/* We want to receive the page-flip event and we want this
+	 * call to be non-blocking.
+	 */
+	flags = DRM_MODE_PAGE_FLIP_EVENT | DRM_MODE_ATOMIC_NONBLOCK;
+        ret = drmModeAtomicCommit(fd, req, flags, dev);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+/*
+ * modeset_draw() changes heavily from all previous examples. The rendering has
+ * moved into another helper modeset_draw_dev() below, but modeset_draw() is now
+ * responsible of controlling when we have to redraw the outputs.
+ *
+ * So what we do: first redraw all outputs. We initialize the r/g/b/_up
+ * variables of each output first, although, you can safely ignore these.
+ * They're solely used to compute the next color. Then we call
+ * modeset_draw_dev() for each output. This function _always_ redraws the output
+ * and schedules a buffer-swap/flip for the next vertical-blank.
+ * We now have to wait for each vertical-blank to happen so we can draw the next
+ * frame. If a vblank happens, we simply call modeset_draw_dev() again and wait
+ * for the next vblank.
+ *
+ * Note: Different monitors can have different refresh-rates. That means, a
+ * vblank event is always assigned to a CRTC. Hence, we get different vblank
+ * events for each CRTC/modeset_dev that we use. This also means, that our
+ * framerate-controlled color-morphing is different on each monitor. If you want
+ * exactly the same frame on all monitors, we would have to share the
+ * color-values between all devices. However, for simplicity reasons, we don't
+ * do this here.
+ *
+ * So the last piece missing is how we get vblank events. libdrm provides
+ * drmWaitVBlank(), however, we aren't interested in _all_ vblanks, but only in
+ * the vblanks for our page-flips. We could use drmWaitVBlank() but there is a
```

Comment:
No, we could not use `drmWaitVBlank()` because it would be completely independent of the pageflip we programmed.